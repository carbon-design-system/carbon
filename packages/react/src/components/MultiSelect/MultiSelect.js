/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { WarningFilled16, WarningAltFilled16 } from '@carbon/icons-react';
import { settings } from 'carbon-components';
import cx from 'classnames';
import Downshift, { useSelect } from 'downshift';
import isEqual from 'lodash.isequal';
import PropTypes from 'prop-types';
import React, { useRef, useState } from 'react';
import ListBox, { PropTypes as ListBoxPropTypes } from '../ListBox';
import { sortingPropTypes } from './MultiSelectPropTypes';
import { defaultItemToString } from './tools/itemToString';
import { defaultSortItems, defaultCompareItems } from './tools/sorting';
import { useSelection } from '../../internal/Selection';
import setupGetInstanceId from '../../tools/setupGetInstanceId';
import { mapDownshiftProps } from '../../tools/createPropAdapter';
import mergeRefs from '../../tools/mergeRefs';
import { keys, match } from '../../internal/keyboard';

const { prefix } = settings;
const noop = () => {};
const getInstanceId = setupGetInstanceId();
const {
  ItemClick,
  MenuBlur,
  MenuKeyDownArrowDown,
  MenuKeyDownArrowUp,
  MenuKeyDownEnter,
  MenuKeyDownEscape,
  MenuKeyDownSpaceButton,
  ToggleButtonClick,
} = useSelect.stateChangeTypes;

const MultiSelect = React.forwardRef(function MultiSelect(
  {
    className: containerClassName,
    id,
    items,
    itemToString,
    titleText,
    helperText,
    label,
    type,
    size,
    disabled,
    initialSelectedItems,
    sortItems,
    compareItems,
    clearSelectionText,
    clearSelectionDescription,
    light,
    invalid,
    invalidText,
    warn,
    warnText,
    useTitleInItem,
    translateWithId,
    downshiftProps,
    open,
    selectionFeedback,
    onChange,
    onMenuChange,
    direction,
  },
  ref
) {
  const { current: multiSelectInstanceId } = useRef(getInstanceId());
  const [highlightedIndex, setHighlightedIndex] = useState(null);
  const [isOpen, setIsOpen] = useState(open);
  const [prevOpenProp, setPrevOpenProp] = useState(open);
  const [topItems, setTopItems] = useState([]);
  const {
    selectedItems: controlledSelectedItems,
    onItemChange,
    clearSelection,
  } = useSelection({
    disabled,
    initialSelectedItems,
    onChange,
  });

  const {
    getToggleButtonProps,
    getLabelProps,
    getMenuProps,
    getItemProps,
    selectedItem: selectedItems,
  } = useSelect(
    mapDownshiftProps({
      ...downshiftProps,
      highlightedIndex,
      isOpen,
      itemToString: (items) => {
        return items.map((item) => itemToString(item)).join(', ');
      },
      onStateChange,
      selectedItem: controlledSelectedItems,
      items,
    })
  );

  /**
   * wrapper function to forward changes to consumer
   */
  const setIsOpenWrapper = (open) => {
    setIsOpen(open);
    if (onMenuChange) {
      onMenuChange(open);
    }
  };

  /**
   * programmatically control this `open` prop
   */
  if (prevOpenProp !== open) {
    setIsOpenWrapper(open);
    setPrevOpenProp(open);
  }

  const inline = type === 'inline';
  const showWarning = !invalid && warn;

  const wrapperClasses = cx(
    `${prefix}--multi-select__wrapper`,
    `${prefix}--list-box__wrapper`,
    {
      [`${prefix}--multi-select__wrapper--inline`]: inline,
      [`${prefix}--list-box__wrapper--inline`]: inline,
      [`${prefix}--multi-select__wrapper--inline--invalid`]: inline && invalid,
      [`${prefix}--list-box__wrapper--inline--invalid`]: inline && invalid,
    }
  );
  const titleClasses = cx(`${prefix}--label`, {
    [`${prefix}--label--disabled`]: disabled,
  });
  const helperId = !helperText
    ? undefined
    : `multiselect-helper-text-${multiSelectInstanceId}`;
  const fieldLabelId = `multiselect-field-label-${multiSelectInstanceId}`;
  const helperClasses = cx(`${prefix}--form__helper-text`, {
    [`${prefix}--form__helper-text--disabled`]: disabled,
  });

  const className = cx(`${prefix}--multi-select`, containerClassName, {
    [`${prefix}--multi-select--invalid`]: invalid,
    [`${prefix}--multi-select--warning`]: showWarning,
    [`${prefix}--multi-select--inline`]: inline,
    [`${prefix}--multi-select--selected`]:
      selectedItems && selectedItems.length > 0,
    [`${prefix}--list-box--up`]: direction === 'top',
  });

  const sortOptions = {
    selectedItems: controlledSelectedItems,
    itemToString,
    compareItems,
    locale: 'en',
  };

  if (selectionFeedback === 'fixed') {
    sortOptions.selectedItems = [];
  } else if (selectionFeedback === 'top-after-reopen') {
    sortOptions.selectedItems = topItems;
  }

  function onStateChange(changes) {
    if (changes.isOpen && !isOpen) {
      setTopItems(controlledSelectedItems);
    }

    const { type } = changes;
    switch (type) {
      case ItemClick:
      case MenuKeyDownSpaceButton:
      case MenuKeyDownEnter:
        onItemChange(changes.selectedItem);
        break;
      case MenuKeyDownArrowDown:
      case MenuKeyDownArrowUp:
        setHighlightedIndex(changes.highlightedIndex);
        break;
      case MenuBlur:
      case MenuKeyDownEscape:
        setIsOpenWrapper(false);
        setHighlightedIndex(changes.highlightedIndex);
        break;
      case ToggleButtonClick:
        setIsOpenWrapper(changes.isOpen || false);
        setHighlightedIndex(changes.highlightedIndex);
        break;
    }
  }

  const onKeyDown = (e) => {
    if (match(e, keys.Delete) && !disabled) {
      clearSelection();
      e.stopPropagation();
    }
  };

  const toggleButtonProps = getToggleButtonProps();

  return (
    <div className={wrapperClasses}>
      <label className={titleClasses} {...getLabelProps()}>
        {titleText && titleText}
        {selectedItems.length > 0 && (
          <span className={`${prefix}--visually-hidden`}>
            {clearSelectionDescription} {selectedItems.length},
            {clearSelectionText}
          </span>
        )}
      </label>
      <ListBox
        type={type}
        size={size}
        className={className}
        disabled={disabled}
        light={light}
        invalid={invalid}
        invalidText={invalidText}
        warn={warn}
        warnText={warnText}
        isOpen={isOpen}
        id={id}>
        {invalid && (
          <WarningFilled16 className={`${prefix}--list-box__invalid-icon`} />
        )}
        {showWarning && (
          <WarningAltFilled16
            className={`${prefix}--list-box__invalid-icon ${prefix}--list-box__invalid-icon--warning`}
          />
        )}
        <button
          type="button"
          className={`${prefix}--list-box__field`}
          disabled={disabled}
          aria-disabled={disabled}
          {...toggleButtonProps}
          ref={mergeRefs(toggleButtonProps.ref, ref)}
          onKeyDown={onKeyDown}>
          {selectedItems.length > 0 && (
            <ListBox.Selection
              clearSelection={!disabled ? clearSelection : noop}
              selectionCount={selectedItems.length}
              translateWithId={translateWithId}
              disabled={disabled}
            />
          )}
          <span id={fieldLabelId} className={`${prefix}--list-box__label`}>
            {label}
          </span>
          <ListBox.MenuIcon isOpen={isOpen} translateWithId={translateWithId} />
        </button>
        <ListBox.Menu aria-multiselectable="true" {...getMenuProps()}>
          {isOpen &&
            sortItems(items, sortOptions).map((item, index) => {
              const itemProps = getItemProps({
                item,
                // we don't want Downshift to set aria-selected for us
                // we also don't want to set 'false' for reader verbosity's sake
                ['aria-selected']: isChecked ? true : null,
              });
              const itemText = itemToString(item);
              const isChecked =
                selectedItems.filter((selected) => isEqual(selected, item))
                  .length > 0;
              return (
                <ListBox.MenuItem
                  key={itemProps.id}
                  isActive={isChecked}
                  aria-label={itemText}
                  isHighlighted={highlightedIndex === index}
                  title={itemText}
                  {...itemProps}>
                  <div className={`${prefix}--checkbox-wrapper`}>
                    <span
                      title={useTitleInItem ? itemText : null}
                      className={`${prefix}--checkbox-label`}
                      data-contained-checkbox-state={isChecked}
                      id={`${itemProps.id}__checkbox`}>
                      {itemText}
                    </span>
                  </div>
                </ListBox.MenuItem>
              );
            })}
        </ListBox.Menu>
      </ListBox>
      {!inline && !invalid && !warn && helperText && (
        <div id={helperId} className={helperClasses}>
          {helperText}
        </div>
      )}
    </div>
  );
});

MultiSelect.displayName = 'MultiSelect';
MultiSelect.propTypes = {
  ...sortingPropTypes,

  /**
   * Specify the text that should be read for screen readers that describes total items selected
   */
  clearSelectionDescription: PropTypes.string,

  /**
   * Specify the text that should be read for screen readers to clear selection.
   */
  clearSelectionText: PropTypes.string,

  /**
   * Specify the direction of the multiselect dropdown. Can be either top or bottom.
   */
  direction: PropTypes.oneOf(['top', 'bottom']),

  /**
   * Disable the control
   */
  disabled: PropTypes.bool,

  /**
   * Additional props passed to Downshift
   */
  downshiftProps: PropTypes.shape(Downshift.propTypes),

  /**
   * Specify a custom `id`
   */
  id: PropTypes.string.isRequired,

  /**
   * Allow users to pass in arbitrary items from their collection that are
   * pre-selected
   */
  initialSelectedItems: PropTypes.array,

  /**
   * Is the current selection invalid?
   */
  invalid: PropTypes.bool,

  /**
   * If invalid, what is the error?
   */
  invalidText: PropTypes.node,

  /**
   * Helper function passed to downshift that allows the library to render a
   * given item to a string label. By default, it extracts the `label` field
   * from a given item to serve as the item label in the list.
   */
  itemToString: PropTypes.func,

  /**
   * We try to stay as generic as possible here to allow individuals to pass
   * in a collection of whatever kind of data structure they prefer
   */
  items: PropTypes.array.isRequired,

  /**
   * Generic `label` that will be used as the textual representation of what
   * this field is for
   */
  label: PropTypes.node.isRequired,

  /**
   * `true` to use the light version.
   */
  light: PropTypes.bool,

  /**
   * Specify the locale of the control. Used for the default `compareItems`
   * used for sorting the list of items in the control.
   */
  locale: PropTypes.string,

  /**
   * `onChange` is a utility for this controlled component to communicate to a
   * consuming component what kind of internal state changes are occuring.
   */
  onChange: PropTypes.func,

  /**
   * `onMenuChange` is a utility for this controlled component to communicate to a
   * consuming component that the menu was opend(`true`)/closed(`false`).
   */
  onMenuChange: PropTypes.func,

  /**
   * Initialize the component with an open(`true`)/closed(`false`) menu.
   */
  open: PropTypes.bool,

  /**
   * Specify feedback (mode) of the selection.
   * `top`: selected item jumps to top
   * `fixed`: selected item stays at it's position
   * `top-after-reopen`: selected item jump to top after reopen dropdown
   */
  selectionFeedback: PropTypes.oneOf(['top', 'fixed', 'top-after-reopen']),

  /**
   * Specify the size of the ListBox. Currently supports either `sm`, `md` or `lg` as an option.
   */
  size: ListBoxPropTypes.ListBoxSize,

  /**
   * Provide text to be used in a `<label>` element that is tied to the
   * multiselect via ARIA attributes.
   */
  titleText: PropTypes.node,

  /**
   * Callback function for translating ListBoxMenuIcon SVG title
   */
  translateWithId: PropTypes.func,

  /**
   * Specify 'inline' to create an inline multi-select.
   */
  type: PropTypes.oneOf(['default', 'inline']),

  /**
   * Specify title to show title on hover
   */
  useTitleInItem: PropTypes.bool,

  /**
   * Specify whether the control is currently in warning state
   */
  warn: PropTypes.bool,

  /**
   * Provide the text that is displayed when the control is in warning state
   */
  warnText: PropTypes.node,
};

MultiSelect.defaultProps = {
  compareItems: defaultCompareItems,
  disabled: false,
  locale: 'en',
  itemToString: defaultItemToString,
  initialSelectedItems: [],
  sortItems: defaultSortItems,
  type: 'default',
  light: false,
  title: false,
  open: false,
  selectionFeedback: 'top-after-reopen',
  direction: 'bottom',
  clearSelectionText: 'To clear selection, press Delete or Backspace,',
  clearSelectionDescription: 'Total items selected: ',
};

export default MultiSelect;
