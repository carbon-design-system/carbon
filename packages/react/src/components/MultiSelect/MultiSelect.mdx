import { Props } from '@storybook/addon-docs/blocks';

# MultiSelect

[Source code](https://github.com/carbon-design-system/carbon/tree/main/packages/react/src/components/MultiSelect)
&nbsp;|&nbsp;
[Usage guidelines](https://www.carbondesignsystem.com/components/dropdown/usage#multiselect)
&nbsp;|&nbsp;
[Accessibility](https://www.carbondesignsystem.com/components/dropdown/accessibility)

## Table of Contents

## Overview

## Component API

<Props />

## Feedback

Help us improve this component by providing feedback, asking questions on Slack,
or updating this file on
[GitHub](https://github.com/carbon-design-system/carbon/edit/main/packages/react/src/components/MultiSelect/MultiSelect.mdx).
