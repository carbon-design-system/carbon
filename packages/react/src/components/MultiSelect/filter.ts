/**
 * Copyright IBM Corp. 2016, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// TODO [@carbon-design-system/monorepo-reviewers]: This file was in the
// `ComboBox` directory before but it wasn't used there. Now it's used in
// `FilterableMultiSelect`. Is that expected?

export const defaultFilterItems = <ItemType>(
  items: ItemType[],
  {
    itemToString,
    inputValue,
  }: {
    itemToString: (item: ItemType | null) => string;
    inputValue: string | null;
  }
): ItemType[] => {
  if (!inputValue) return items;

  const normalizedInput = inputValue.toLowerCase();

  return items.filter((item) =>
    itemToString(item).toLowerCase().includes(normalizedInput)
  );
};
