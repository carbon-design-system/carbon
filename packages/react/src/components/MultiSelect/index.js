/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as FeatureFlags from '@carbon/feature-flags';
import { deprecateFieldOnObject } from '../../internal/deprecateFieldOnObject';

import MultiSelect from './MultiSelect';
import { default as FilterableMultiSelectClassic } from './FilterableMultiSelect';
import { default as FilterableMultiSelectNext } from './next/FilterableMultiSelect';

FilterableMultiSelectNext.displayName = 'MultiSelect.Filterable';
MultiSelect.Filterable = FilterableMultiSelectClassic;

export const FilterableMultiSelect = FeatureFlags.enabled('enable-v11-release')
  ? FilterableMultiSelectNext
  : FilterableMultiSelectClassic;

if (__DEV__) {
  deprecateFieldOnObject(MultiSelect, 'Filterable', FilterableMultiSelect);
}

export default MultiSelect;
