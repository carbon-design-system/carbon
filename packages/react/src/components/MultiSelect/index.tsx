/**
 * Copyright IBM Corp. 2016, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import MultiSelect from './MultiSelect';
import { default as FilterableMultiSelect } from './FilterableMultiSelect';

FilterableMultiSelect.displayName = 'FilterableMultiSelect';

export { FilterableMultiSelect, MultiSelect };
export default MultiSelect;
