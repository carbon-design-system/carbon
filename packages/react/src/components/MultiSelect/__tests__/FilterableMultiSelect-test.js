/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import MultiSelect from '../../MultiSelect';
import {
  assertMenuOpen,
  assertMenuClosed,
  findMenuIconNode,
  openMenu,
  generateItems,
  generateGenericItem,
} from '../../ListBox/test-helpers';

const listItemName = 'ListBoxMenuItem';

describe('MultiSelect.Filterable', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      id: 'test-filterable-multiselect',
      disabled: false,
      items: generateItems(5, generateGenericItem),
      initialSelectedItems: [],
      onChange: jest.fn(),
      onMenuChange: jest.fn(),
      placeholder: 'Placeholder...',
    };
  });

  it('should render', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display all items when the menu is open initially', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    openMenu(wrapper);
    expect(wrapper.find(listItemName).length).toBe(mockProps.items.length);
  });

  it('should initially have the menu open when open prop is provided', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} open />);
    assertMenuOpen(wrapper, mockProps);
  });

  it('should open the menu with a down arrow', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    const menuIconNode = findMenuIconNode(wrapper);

    menuIconNode.simulate('keyDown', { key: 'ArrowDown' });
    assertMenuOpen(wrapper, mockProps);
  });

  it('should let the user toggle the menu by the menu icon', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    findMenuIconNode(wrapper).simulate('click');
    assertMenuOpen(wrapper, mockProps);
    findMenuIconNode(wrapper).simulate('click');
    assertMenuClosed(wrapper);
  });

  it('should not close the menu after a user makes a selection', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    openMenu(wrapper);

    const firstListItem = wrapper.find(listItemName).at(0);

    firstListItem.simulate('click');
    assertMenuOpen(wrapper, mockProps);
  });

  it('should filter a list of items by the input value', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    openMenu(wrapper);
    expect(wrapper.find(listItemName).length).toBe(mockProps.items.length);

    wrapper
      .find('[placeholder="Placeholder..."]')
      .at(1)
      .simulate('change', { target: { value: '3' } });

    expect(wrapper.find(listItemName).length).toBe(1);
  });

  it('should call `onChange` with each update to selected items', () => {
    const wrapper = mount(
      <MultiSelect.Filterable {...mockProps} selectionFeedback="top" />
    );
    openMenu(wrapper);

    // Select the first two items
    wrapper.find(listItemName).at(0).simulate('click');

    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [mockProps.items[0]],
    });

    wrapper.find(listItemName).at(1).simulate('click');

    expect(mockProps.onChange).toHaveBeenCalledTimes(2);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [mockProps.items[0], mockProps.items[1]],
    });

    // Un-select the next two items
    wrapper.find(listItemName).at(0).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(3);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [mockProps.items[0]],
    });

    wrapper.find(listItemName).at(0).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(4);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [],
    });
  });

  it('should let items stay at thier position after selecting', () => {
    const wrapper = mount(
      <MultiSelect.Filterable {...mockProps} selectionFeedback="fixed" />
    );
    openMenu(wrapper);

    // Select the first two items
    wrapper.find(listItemName).at(1).simulate('click');

    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [mockProps.items[1]],
    });

    wrapper.find(listItemName).at(1).simulate('click');

    expect(mockProps.onChange).toHaveBeenCalledTimes(2);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItems: [],
    });
  });

  it('should not clear input value after a user makes a selection', () => {
    const wrapper = mount(<MultiSelect.Filterable {...mockProps} />);
    openMenu(wrapper);

    wrapper
      .find('[placeholder="Placeholder..."]')
      .at(1)
      .simulate('change', { target: { value: '3' } });

    wrapper.find(listItemName).at(0).simulate('click');

    expect(
      wrapper.find('[placeholder="Placeholder..."]').at(1).props().value
    ).toEqual('3');
  });
});
