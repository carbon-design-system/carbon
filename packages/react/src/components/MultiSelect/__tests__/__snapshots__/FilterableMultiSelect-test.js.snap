// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect.Filterable should render 1`] = `
<MultiSelect.Filterable
  ariaLabel="Choose an item"
  compareItems={[Function]}
  disabled={false}
  filterItems={[Function]}
  id="test-filterable-multiselect"
  initialSelectedItems={Array []}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  light={false}
  locale="en"
  onChange={[MockFunction]}
  onMenuChange={[MockFunction]}
  open={false}
  placeholder="Placeholder..."
  selectionFeedback="top-after-reopen"
  sortItems={[Function]}
>
  <div
    className="bx--multi-select__wrapper bx--list-box__wrapper"
  >
    <Selection
      disabled={false}
      initialSelectedItems={Array []}
      onChange={[Function]}
      render={[Function]}
    >
      <Downshift
        defaultHighlightedIndex={null}
        defaultIsOpen={false}
        environment={[Window]}
        getA11yStatusMessage={[Function]}
        id="test-filterable-multiselect"
        inputValue=""
        isOpen={false}
        itemToString={[Function]}
        onChange={[Function]}
        onInputValueChange={[Function]}
        onOuterClick={[Function]}
        onSelect={[Function]}
        onStateChange={[Function]}
        onUserAction={[Function]}
        scrollIntoView={[Function]}
        selectedItem={Array []}
        selectedItemChanged={[Function]}
        stateReducer={[Function]}
        suppressRefError={false}
      >
        <ListBox
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-labelledby="test-filterable-multiselect-label"
          aria-owns={null}
          className="bx--multi-select bx--combo-box bx--multi-select--filterable"
          disabled={false}
          isOpen={false}
          light={false}
          role="combobox"
          type="default"
        >
          <div
            aria-expanded={false}
            aria-haspopup="listbox"
            aria-labelledby="test-filterable-multiselect-label"
            aria-owns={null}
            className="bx--multi-select bx--combo-box bx--multi-select--filterable bx--list-box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="combobox"
          >
            <ListBoxField
              aria-haspopup={true}
              aria-labelledby="test-filterable-multiselect-label"
              data-toggle={true}
              disabled={false}
              id="test-filterable-multiselect"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              role="button"
              type="button"
            >
              <div
                aria-haspopup={true}
                aria-labelledby="test-filterable-multiselect-label"
                className="bx--list-box__field"
                data-toggle={true}
                id="test-filterable-multiselect"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                role="button"
                tabIndex={-1}
                type="button"
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-controls={null}
                  aria-labelledby="test-filterable-multiselect-label"
                  autoComplete="off"
                  className="bx--text-input bx--text-input--empty"
                  disabled={false}
                  id="test-filterable-multiselect-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Placeholder..."
                  value=""
                />
                <ListBoxMenuIcon
                  isOpen={false}
                  translateWithId={[Function]}
                >
                  <div
                    className="bx--list-box__menu-icon"
                  >
                    <ForwardRef(ChevronDown16)
                      aria-label="Open menu"
                      name="chevron--down"
                    >
                      <Icon
                        aria-label="Open menu"
                        fill="currentColor"
                        height={16}
                        name="chevron--down"
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 16 16"
                        width={16}
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <svg
                          aria-label="Open menu"
                          fill="currentColor"
                          focusable="false"
                          height={16}
                          name="chevron--down"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 16 16"
                          width={16}
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                          />
                          <title>
                            Open menu
                          </title>
                        </svg>
                      </Icon>
                    </ForwardRef(ChevronDown16)>
                  </div>
                </ListBoxMenuIcon>
              </div>
            </ListBoxField>
          </div>
        </ListBox>
      </Downshift>
    </Selection>
  </div>
</MultiSelect.Filterable>
`;
